### ISSUE TITLE
Redis cluster errors under high load

### ISSUE THREAD

**Issue by crodwell** [go-redis/redis#695]
Hi, I'm connecting to a redis cluster and after a bunch of heavy queries.. hundreds of sequential HGETs + HSETs, I get an error eventually doing an HSET
*errors.errorString)(0xc42017e730)(redis: client is closed)

Any idea what could be the cause?

```
Cc = redis.NewClusterClient(&redis.ClusterOptions{
    Addrs:     []string{
	    	redis_host + ":7000", 
	    	redis_host + ":7001", 
	    	redis_host + ":7002", 
	    	redis_host + ":7003", 
	    	redis_host + ":7004", 
	    	redis_host + ":7005"},
	   	MaxRedirects: 16,
 		ReadTimeout: time.Duration(200) * time.Millisecond,
 		WriteTimeout: time.Duration(200) * time.Millisecond,
		PoolSize: 10,
		PoolTimeout: time.Duration(15000) * time.Millisecond,
		IdleTimeout: time.Duration(15000) * time.Millisecond,
		ReadOnly: true,
		RouteByLatency: true,
	})
```

**Comment by vmihailenco** [comment]
You somewhere close the client. grep for `Close` and inspect every one of them.

**Comment by vmihailenco** [comment]
Another possibility is that:
- you have really loaded application and requests/goroutines are queued for more than 5 seconds
- Redis cluster is reconfigured

In that case Redis cluster client can close/remove unused cluster nodes (because cluster is reconfigured), but some goroutines can still use old nodes (because goroutines were queued/blocked for some time). If you are interested in the code you can start from [here](https://github.com/go-redis/redis/blob/master/cluster.go#L841-L856). The simple (and also proper) fix would be to retry commands on `redis.ErrClosed` error.

To confirm that my theory is true can you check if following commands are processed properly? I.e. you should occacionally see `ErrClosed` errors, but then Redis Cluster client should recover and following commands are processed normally.

**Comment by vmihailenco** [comment]
>When you say redis cluster reconfigured, you mean it's being changed after connection right? in which case no, that's not happening

Yes, new node is added or existing is removed after `NewCluster` is called.

>It made me think the Cluster is not releasing the pool when done

You need to call `cluster.Close` to release connections in the pool.

> they may pass first time but after subsequent runs it will fail on high query tests (hundreds, not thousands mind you).

If you have unit tests that fail you can insert `fmt.Println` at https://github.com/go-redis/redis/blob/master/cluster.go#L248 to check if node is really closed by cluster client. `fmt.Println` at https://github.com/go-redis/redis/blob/master/cluster.go#L862 should tell you if cluster state is reloaded during tests.

**Comment by vmihailenco** [comment]
Please try fix release in v6.8.2.

**Comment by crodwell** [comment]
My setup has the cluster defined in it's own package 'redisclient' and connect happens on API start then it is called from many goroutines in parallel going forward, I thought this was the intention by the docs:

> Close closes the cluster client, releasing any open resources. It is rare to Close a ClusterClient, as the ClusterClient is meant to be long-lived and shared between many goroutines.

I'll check your debug advice and report back soon.
